### DISCUSSION TITLE
acceleration plot behaviour differences

### DISCUSSION THREAD

**Discussion by chansigit** [aristoteleo/dynamo-release#365] [Q&A] [answered]
I tried several types of acceleration plots and found their behaviours puzzling. 

# 0.heatmap
 `sc.pl.heatmap(adata, markers, groupby='field_louvain', swap_axes=True, layer='acceleration', dendrogram=True)`
![image](https://user-images.githubusercontent.com/18084613/170995113-f26a8afe-a454-4495-978b-d13fa490e3d3.png)
The heatmap looks fine to me

# 1. sc.pl.umap
`sc.pl.umap(adata, layer='acceleration', color='Cndp2')`

<img width="602" alt="image" src="https://user-images.githubusercontent.com/18084613/170994511-495b5629-010d-4901-bc04-c7ae81a99b6f.png">

It also looks fine to me as it is consistent with the heatmap

---

# 2. dyn.pl.umap
 `dyn.pl.umap(adata, layer='acceleration', color='Cndp2', pointsize=0.05, alpha=1,frontier=False)`

<img width="504" alt="image" src="https://user-images.githubusercontent.com/18084613/170994641-f004caeb-ee54-4df3-80c4-ca98287d97e2.png">

I thought this plot would grab the embedding from the "X_umap" but it probably grab embeddings from the "acceleration_umap"
Question 1: What is "acceleration_umap"?  Can I use "X_umap" with this function?

---

# 3. dyn.pl.acceleration

`dyn.pl.acceleration(adata, basis='umap', color='Cndp2', pointsize=0.05, alpha=1,frontier=False)`

<img width="889" alt="image" src="https://user-images.githubusercontent.com/18084613/170994679-2a4a3055-ce5c-446c-8bb6-20b009c4ecf6.png">

This plot looks weird to me because the colorbar ranges from 0 to 0.4 . 
But the previous heatmap and plots told me the acceleration ranges from 0 to 0.0002 
Question 2: Why did I got acceleration values of quite different scales from different plots?

---

# 4. another dyn.pl.acceleration

 `dyn.pl.acceleration(adata, basis='pca', color='Cndp2', pointsize=0.05, alpha=1,frontier=False)`

<img width="920" alt="image" src="https://user-images.githubusercontent.com/18084613/170994801-be893214-8fa6-44d9-9eea-d0c00f10a0d1.png">

This plot looks weird to me too because the colorbar ranges from 0 to 0.03 when I changed the basis to 'pca' . 
But the previous heatmap and plots told me the acceleration ranges from 0 to 0.0002 
Question 3: Does the `basis` parameter here changes the embedding coordinates or the actual acceleration values computed with different methods?

---

my anndata:
```
AnnData object with n_obs × n_vars = 4284 × 31053
    obs: 'Clusters', 'ann0608', 'phase', 'nGenes', 'nCounts', 'pMito', 'pass_basic_filter', 'unspliced_Size_Factor', 'initial_unspliced_cell_size', 'spliced_Size_Factor', 'initial_spliced_cell_size', 'Size_Factor', 'initial_cell_size', 'ntr', 'cell_cycle_phase', 'jaccard_velocity_confidence', 'umap_ddhodge_div', 'umap_ddhodge_potential', 'curl_umap', 'divergence_umap', 'control_point_umap', 'inlier_prob_umap', 'obs_vf_angle_umap', 'control_point_pca', 'inlier_prob_pca', 'obs_vf_angle_pca', 'speed_pca', 'divergence_pca', 'acceleration_pca', 'curvature_pca', 'pca_ddhodge_div', 'pca_ddhodge_potential', 'speed_umap', 'acceleration_umap', 'curvature_umap', 'field_louvain', 'field_leiden'
    var: 'Accession', 'Chromosome', 'End', 'Start', 'Strand', 'nCells', 'nCounts', 'pass_basic_filter', 'log_m', 'log_cv', 'score', 'frac', 'use_for_pca', 'ntr', 'beta', 'gamma', 'half_life', 'alpha_b', 'alpha_r2', 'gamma_b', 'gamma_r2', 'gamma_logLL', 'delta_b', 'delta_r2', 'bs', 'bf', 'uu0', 'ul0', 'su0', 'sl0', 'U0', 'S0', 'total0', 'use_for_dynamics', 'use_for_transition', 'avg_prog_confidence', 'avg_mature_confidence', 'avg_confidence', 'confident_genes'
    uns: 'ann0608_colors', 'pp', 'velocyto_SVR', 'PCs', 'explained_variance_ratio_', 'pca_mean', 'pca_fit', 'feature_selection', 'cell_phase_genes', 'dynamics', 'neighbors', 'grid_velocity_umap', 'gene_wise_confidence', 'VecFld_umap', 'grid_velocity_pca', 'VecFld_pca', 'field_louvain', 'field_louvain_colors', 'field_leiden', 'field_leiden_colors', 'rank_velocity_S', 'rank_abs_velocity_S'
    obsm: 'X_pca', 'X_umap', 'X_umap_rna', 'X', 'cell_cycle_scores', 'velocity_umap', 'velocity_umap_SparseVFC', 'X_umap_SparseVFC', 'velocity_pca', 'velocity_pca_SparseVFC', 'X_pca_SparseVFC', 'acceleration_pca', 'curvature_pca', 'acceleration_umap', 'curvature_umap'
    layers: 'ambiguous', 'matrix', 'spliced', 'unspliced', 'X_unspliced', 'X_spliced', 'M_u', 'M_uu', 'M_s', 'M_us', 'M_ss', 'velocity_S', 'acceleration', 'curvature'
    obsp: 'moments_con', 'distances', 'connectivities', 'pearson_transition_matrix', 'umap_ddhodge', 'pca_ddhodge', 'vf_feature_knn'
```

Thank you for your answers!

**Comment by Xiaojieqiu** [comment, marked as answer]
I am sorry so I cannot really answer each question closely in the near future because from now on I need to prepare my job interviews and other stuff stuff these days. My general answer is that there is a risk to mix tools together and you tend to use other tool's design principle over dynamo which is at times not reasonable. 

1.  the sc heatmap and umap should really use a divergent colormap (such as `bwr` instead of `viridis`) because acceleation values can be negative 

2.` dyn.pl.umap(adata, color=['Cndp2'], layer='acceleration', frontier=True)` uses `acceleration_umap` (as it is shown from the running message) that is because you also calculate `acceleration` magnitude for the umap embedding. if you delete or change `adata.obsm['acceleration_umap']`,  it will use the umap. I will think the best way to avoid using this later but probably the running message is already enough. Note that `acceleration_umap` is the acceleration magnitude for the acceleration vectors on umap space. I have explained this to you in another github issue. 

3. `dyn.pl.acceleration` is used to plot the acceleration magnitude (`not the acceleration value`) for each cell

**Reply by chansigit** [reply]
Thank you so much Xiaojie. Your agile responses really impressed me a lot. Wish everything goes smoothly to you!
I now realized I messed the value and the magnitude.
